libs/tlv-derive/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+//! This crate provides procedural macros for the `tlv` crate. The macros are
+//! used to derive the `Encode` and `Decode` traits for structs. The `Encode`
+//! trait is used to encode a struct into a TLV format. The `Decode` trait is
+//! used to decode a struct from a TLV format.
+
+use proc_macro::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, spanned::Spanned, DeriveInput};
+
+/// Generates an implementation of the `Encode` trait for the annotated struct.
+/// The implementation will encode the struct into a TLV format.
+#[proc_macro_derive(Encode, attributes(tlv))]
+pub fn encode_derive(input: TokenStream) -> TokenStream {
+    // Construct a representation of Rust code as a syntax tree
+    // that we can manipulate
+    let DeriveInput { ident, data, .. } = parse_macro_input!(input);
+
+    let name = ident;
+    let entities = match data {
+        syn::Data::Struct(data) => match data.fields {
+            syn::Fields::Named(fields) => fields
+                .named
+                .into_iter()
+                .map(|f| parse_field_tlv_entity(&f).unwrap())
+                .collect::<Vec<_>>(),
+            _ => panic!("only named fields are supported"),
+        },
+        _ => panic!("only structs are supported"),
+    };
+    let tags = entities
+        .iter()
+        .map(|e| match e.tag {
+            tlv::Tag::U8(tag) => quote! { tlv::Tag::U8(#tag) },
+            tlv::Tag::U16(tag) => quote! { tlv::Tag::U16(#tag) },
+        })
+        .collect::<Vec<_>>();
+    let idents = entities.iter().map(|e| &e.ident).collect::<Vec<_>>();
+    let gen = quote! {
```

Comment:
The `quote!` macro allows building token streams with interpolation of syntax nodes.